### PR TITLE
Matt/unit test update

### PR DESCRIPTION
**Pull request by matt001k**
This is a PR to see if you like the reverse module control structure,
the architecture will feed as so
```ASCII
--------------------------
|         bcmp module           |
--------------------------
                     |
--------------------------
|   bcmp message types   |
--------------------------
                    |
--------------------------
|        packet module         |
--------------------------
```

I tried my best to keep it clean.
If you like this stuff, I will also use this PR to get rid of all of the common items that @victorsowa12 placed in `bm_common`

### CODE REVIEW

File: test/src/packet_test.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        .cfg =
+            {
+                sizeof(BCMPHeartbeat),
+                false,
+                NULL,
+                bcmp_process_heartbeat,
+            },
+        .ll_item = {0},
+    };
+    ASSERT_EQ(packet_add(&heartbeat_packet, BCMPHeartbeatMessage), BmOK);
```
Comment (matt001k) [inline test/src/packet_test.cpp:100]: This shows how to add the configuration struct to the packet module.

File: src/ll.c [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+ */
+BmErr ll_traverse(LL *ll, LLTraverseCb cb, void *arg) {
+  BmErr ret = BmEINVAL;
+  void *data = NULL;
+
+  if (ll && cb) {
+    ll_reset_cursor(ll);
+    do {
+      ret = ll_advance_cursor(ll, &data);
+      if (ret == 0) {
```
Comment (victorsowa12) [inline src/ll.c:184]: lets compare against BmOk

File: src/packet.c [hunk 3]
```diff
@@ -694,36 +424,32 @@ BmErr parse(void *payload) {
...
-          if (PACKET_LUT[i].parse && (err = PACKET_LUT[i].parse(data)) != BmOK) {
-            printf("Error completing parse cb: 0x%X\n", err);
-          } else {
-            break;
-          }
+      if (cb) {
+        // If message is a reply, utilize associated cb
+        err = cb(data.payload);
+      } else {
+        // Utilize LUT parsing callback
```
Comment (victorsowa12) [inline src/packet.c:450]: nit: not using LUT anymore

**Review by victorsowa12 — APPROVED** [review]
Couple small things, but otherwise looks good to me

**Review by towynlin — CHANGES_REQUESTED** [review]
(1) How well does this design reduce the need for include guards?

Very well, but not entirely, which now seems impossible. All translation units other than the most simple end up including util.h multiple times, so util.h needs guards. For example, heartbeat.c includes it 3 times.

    heartbeat.c:
    - heartbeat.h: util.h
    - packet.h:
      - ll.h: util.h
      - messages.h
      - util.h

The linked list header ll.h currently has guards even though it doesn't need them. It's easy to imagine they might be needed though. It's also easy to imagine that anything in bm_common might need include guards. Perhaps we should enforce them there and can avoid them in other repos? Probably not. Maybe we should just add them. 🤷‍♂️ 

(2) How clearly understandable is the code?

I feel a bit dense, but I'm really having trouble understanding what's going on. Some docs would be helpful.

The non-testing code does not build, which seems basic to me. The first failure I get is that packet.c includes bm_rtos.h which doesn't exist. We also need a toolchain file. The bm_netif repo has an example.

Just talking through my process of trying to understand... bcmp.h only declares bcmp_init, defined in bcmp.c. Presumably this would be called from outside the repo or if it's internal then from somewhere close to the external shell of the API.

Right now bcmp.c includes heartbeat.h, and bcmp_init calls bcmp_heartbeat_init. This is not declared in heartbeat.h which only has heartbeat_init. Need to make the names match. Declaring the fake in the test hides this failure. I don't think it should be necessary to declare the fake.

Presumably, we would replace the static look-up table struct with a sequence of function calls in `bcmp_init` for each message. Each of those `{message}_init` functions would call `packaet_add`. The thing it has to pass to `packet_add` is a `BCMPPacketItem` which forces the caller to know internal implementation details of how packets are added -- the fact that it will store them as a linked list. I could see passing a `BCMPPacketCfg` instead, and letting `packet_add` construct the linked list item. I think we can improve some of the naming there as well.

**Comment by matt001k** [comment]
@towynlin your understanding is correct.
I addressed the requested changes:

- The packet header file is not reliant on ll.h and no other module have to know about this now
- The program now will build based on a toolchain file (arm-none-eabi-gcc)

Please note the TODOs in the top level CMake file,
I set up a structure to be able to pull in bm_common (and in the future bm requirements) utilizing the fetch content CMake API.
It is pulling the main branch as of now, but in the future we can get this going with custom configurations (what version to pull, etc,)

My next tasks will be:
- Documenting this flow of control
- Potentially renaming things so they make further sense and so they match our contribution requirements (see (this) (https://sofarocean.slack.com/archives/C035EELLPMJ/p1726608239839269)
- Moving over ll to bm_common? (I belive it would live nicely there @victorsowa12 and I saw many spots in bm_protocol that could benefit from using a standard ll library as there is copy pasta all over the place rn :))

**Pull request merged**